solidity/contracts/OngoingAirdrops.sol

```diff
@@ -37,7 +37,39 @@ contract OngoingAirdrops is AccessControl, IOngoingAirdrops {
     bytes32 _campaign,
     bytes32 _root,
     TokenAmount[] calldata _tokensAllocation
-  ) external override {}
+  ) external override onlyRole(ADMIN_ROLE) {
+    if (_campaign == bytes32(0)) revert InvalidCampaign();
+    if (_root == bytes32(0)) revert InvalidMerkleRoot();
+    if (_tokensAllocation.length == 0) revert InvalidTokenAmount();
+
+    for (uint256 i = 0; i < _tokensAllocation.length; i++) {
+      // Build our unique ID for campaign and token address.
```

Comment:
It's usually cheaper if you load the TokenAmount struct into memory and only read the array once

Reply:
I was completely skeptic about this, I really thought `calldata` access was cheaper that `memory`. Wrong, once again 😬 .